GcExtensions.md

```diff
@@ -0,0 +1,429 @@
+the sweep function will not be called on this particular object. This is
+to avoid unnecessary overhead if not all objects of that type require
+extra sweep phase semantics. This function should be called at most once
+per object; if called multiple times, the sweep function may be invoked
+more than once on the given object.
+
+```
+JL_DLLEXPORT void jl_gc_schedule_foreign_sweepfunc(jl_ptls_t ptls,
+        jl_value_t *obj);
+```
+
+### Allocating foreign objects
+
+On the C side, such objects can be allocated using the call
+`jl_gc_alloc_typed()`; the function takes a pointer to the thread's
+thread-local storage, the desired size, and the foreign datatype as its
+arguments.
+
+```
+JL_DLLEXPORT void * jl_gc_alloc_typed(jl_ptls_t ptls, size_t sz,
+  void *ty);
+```
+
+## Conservative scanning
+
+Some external modules may require conservative scanning, especially
+of the stack. This was the case, for example, with our application
+involving the GAP computer algebra system.
+
+We note that conservative scanning should be avoided if at all possible;
+it is not intended as a way to avoid tracking Julia references (for which
+the `root_scanner` callback and custom marking functions offer efficient
+options if other approaches fail), but as a feature of last resort if
+integrating an existing codebase through other means is not viable.
+
+Conservative scanning must be enabled through a call to the following
+function:
+
+```
+void jl_gc_enable_conservative_scanning(void);
+```
+
+This function can be called from C code both before and after `jl_init()` 
+and is thread-safe. Enabling this introduces a very small, but non-zero
+overhead, which is why it is not enabled by default.
+
+In order to handle conservative scanning, we need to expose the fact
+that Julia distinguishes between objects it manages itself (which we
+call "internal objects" in this document) and objects that it manages
+via "malloc()" or similar calls (these we call "external objects").
+
+The proposed functionality relies on calls to Julia to determine
+if a pointer is a reference to an internal object, but leaves it up
+to the author of the foreign code to determine this for external
+objects; to this end, we provide callbacks to notify foreign code
+of the allocation or deallocation of such objects.
+
+The accompanying function pointer types are:
+
+```
+typedef void (*jl_gc_cb_notify_external_alloc_t)(void *addr, size_t size);
+typedef void (*jl_gc_cb_notify_external_free_t)(void *addr);
+```
+
+The allocation callback is invoked with the address and size of the
+new object, the deallocation callback is invoked with the address of
+the object about to be freed. Allocation and deallocation is still
+managed by Julia. The intent here is that foreign code can track
+allocations and deallocations in a data structure of its own if
+needed. An example of this can be seen in the `gcext` test, where
+we use a balanced tree to track allocations.
```

Comment:
It might be nice just to move this logic into Julia and drop these hooks. Most of this tree already exists anyways (since we need it for pool allocations), and I think it might turn out to be useful to have. Anyways, that change can be done at any time.

Reply:
To clarify my understanding of what you wrote here: these notifications are for tracking malloc'ed memory ranges; in particular, to allow to take an arbitrary address and find out if it is inside a memory range we malloc'ed (and to find the start address of that range).  Are you saying this is something you might want to do inside Julia itself? If course we wouldn't mind that, as it'd simplify our code on the other side. We just didn't expect you to like the idea, for performance reasons (the bookkeeping is cheap but not free)